src/rws_subscription.cpp

```diff
@@ -57,11 +62,14 @@ namespace abb :: rws
     // Wait for (non-ping) WebSocket frames.
     do
     {
+      auto last_frame = std::chrono::steady_clock::now();
       now = std::chrono::steady_clock::now();
       if (now >= deadline)
-        BOOST_THROW_EXCEPTION(TimeoutError {"WebSocket frame receive timeout"});
+        BOOST_THROW_EXCEPTION(TimeoutError {"WebSocket Failed to receive new subscription message in " +
+                                            std::to_string(new_message_timeout.count()) + " microseconds."});
 
-      webSocket_.setReceiveTimeout(std::chrono::duration_cast<std::chrono::microseconds>(deadline - now).count());
+      // Set the timeout for the next receive operation (we receive ping-pong messages every 30 seconds from controller).
+      webSocket_.setReceiveTimeout({std::min(ping_pong_timeout.count(), std::chrono::duration_cast<std::chrono::microseconds>(deadline-now).count())});
```

Comment:
What is the meaning of `Timespan` passed to the `setReceiveTimeout()`? I looked at https://docs.pocoproject.org/current/Poco.Net.Socket.html#27682 but couldn't find the answer.

Reply:
It sets the timeout for receiving a message on websocket with function `webSocket_.receiveFrame()`

The implementation is as folllows:
The function `receiveFrame()` waits for the message to arrive for the timeout period passed by `setReceiveTimeout` and then waits again for the actual timeout.

It is somewhat a strange implementation of timeout, but the `webSocket_.receiveFrame()` is blocking for 2x the `Timespan` passed to `setReceiveTimeout`

Reply:
Ah, sorry, my bad. I was mislead by `()` and `{}` and interpreted the line as implicitly calling `Timespan()` constructor with two arguments. Please reformat the code s.t. it is more apparent what is going on in this line. If you need to use `Timespan`, please call the constructor explicitly, because otherwise it is not immediately clear what does `{}` mean.

So you create a `Timespan` which is the min of `ping_pong_timeout` and `deadline-now`, correct?

Reply:
Yes, that is correct. I think I will create a local variable there. It will be more clear

Reply:
> The function `receiveFrame()` waits for the message to arrive for the timeout period passed by `setReceiveTimeout` and then waits again for the actual timeout.
> 
> It is somewhat a strange implementation of timeout, but the `webSocket_.receiveFrame()` is blocking for 2x the `Timespan` passed to `setReceiveTimeout`

This I don't understand. I initially though that it is an overloaded `setReceiveTimeout()`, but it is the same (and the only) `void setReceiveTimeout(const Poco::Timespan& timeout);` as before. [The documentation](https://docs.pocoproject.org/current/Poco.Net.Socket.html#27682) says "Sets the send timeout for the socket.". Why "2x the Timespan"?

Reply:
> > The function `receiveFrame()` waits for the message to arrive for the timeout period passed by `setReceiveTimeout` and then waits again for the actual timeout.
> > It is somewhat a strange implementation of timeout, but the `webSocket_.receiveFrame()` is blocking for 2x the `Timespan` passed to `setReceiveTimeout`
> 
> This I don't understand. I initially though that it is an overloaded `setReceiveTimeout()`, but it is the same (and the only) `void setReceiveTimeout(const Poco::Timespan& timeout);` as before. [The documentation](https://docs.pocoproject.org/current/Poco.Net.Socket.html#27682) says "Sets the send timeout for the socket.". Why "2x the Timespan"?

I have tested it and this is how the function behaves. That's why I had those stamps when throwing exception before.
I spoke whth a collegue from automotive and he told me that this is the way they implement timeouts

Reply:
What the **ck!